README.md
# PivotalToPdf::Formatters

The gem provides the formatters for the pivotal_to_pdf gem.

## Installation

Add this line to your application's Gemfile:

    gem 'pivotal_to_pdf-formatters'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install pivotal_to_pdf-formatters

## Usage

The gem needs to use with pivotal_to_pdf gem. It provides the formatters for the pivotal_to_pdf.

To use an existing formatter, Please see the [pivotal_to_pdf README](https://github.com/ywen/pivotal_to_pdf) **Formatters** section.

To add a formatter, please fork this project, add you own formatter, submit a PR, and I will merge it if I decide no harm can be done and update the version. Once the new version of this gem is released, you can update the gem and your formatter should be ready for use.

### Coding standard

A formatter must have 2 public interfaces for it to work:

- ```initialize``` takes an array of stories
- ```write_to``` method writes to the your chosen output (PDF, HTML, etc.) in your style.

Your formatter can extends from ```PivotalToPdf::Formatters::Base``` to have ```initialize``` method available.

The ```Default``` formatter is a bad example on how to write code. I expect some tests along with your formatter.
Design it good, do not afraid adding more help classes when needed.

##Contributors
* [Yi Wen](https://github.com/ywen)
* [Luke brown](https://github.com/tsdbrown)
* [Magne Land](https://github.com/magneland)

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG.md
v 0.2
  * Add LargeText support
  ```yaml
  formatter: PivotalToPdf::Formatters::LargeText
  ```

v 0.0.4
  * Add Asian font support as a new formatter Asian
  ```yaml
  formatter: PivotalToPdf::Formatters::Asian
  ```



spec/spec_helper.rb
require 'simplecov'
SimpleCov.start do
  SimpleCov.at_exit do
    threshold = 100
    SimpleCov.result.format!
    if SimpleCov.result.covered_percent != threshold
      actual_coverage = SimpleCov.result.covered_percent
      puts "The coverage is #{actual_coverage}%. It has to be 100%"
      exit 1
    end
  end
end
require 'spork'
require 'guard/spork'

Spork.prefork do
  require 'rspec/core'
  Dir["#{File.dirname(__FILE__)}/support/**/*.rb"].each {|f| require f}
  require "#{File.dirname(__FILE__)}/../lib/pivotal_to_pdf-formatters"

  RSpec.configure do |config|
    config.mock_with :rspec
  end
end

Spork.each_run do
  load "#{File.dirname(__FILE__)}/../lib/pivotal_to_pdf-formatters.rb"
  Dir["#{File.dirname(__FILE__)}/../lib/pivotal_to_pdf-formatters/*.rb"].each {|f| load f}
end



lib/pivotal_to_pdf-formatters/default.rb
require 'rubygems'
require 'prawn'
require 'rainbow'

module PivotalToPdf
  module Formatters
    class Default < Base
      def write_to(destination)
        Prawn::Document.generate("#{destination}.pdf",
                                 :page_layout => :landscape,
                                 :margin => [25, 25, 50, 25],
                                 :page_size => [302, 432]) do |pdf|

          setup_font(pdf)

          stories.each_with_index do |story, index|
            padding = 10
            pdf.stroke_color = story.story_color
            pdf.stroke_bounds
            width = 370
            pdf.line_width=6
            # --- Write content
            pdf.bounding_box [pdf.bounds.left+padding, pdf.bounds.top-padding], :width => width do
              pdf.text story.formatted_name, :size => 14, :inline_format => true
              pdf.fill_color "52D017"
              pdf.text story.formatted_labels, :size => 8
              pdf.text "\n", :size => 14
              pdf.fill_color "444444"
              pdf.text story.formatted_description || "", :size => 10, :inline_format => true
              pdf.fill_color "000000"
            end
            pdf.line(pdf.bounds.bottom_left, pdf.bounds.bottom_right)
            pdf.stroke_bounds

            pdf.text_box story.points, :size => 12, :at => [12, 50], :width => width-18 unless story.points.nil?
            pdf.text_box "Owner: " + (story.respond_to?(:owned_by) ? story.owned_by : "None"),
                         :size => 8, :at => [12, 18], :width => width-18

            pdf.fill_color "999999"
            pdf.text_box story.story_type.capitalize, :size => 8, :align => :right, :at => [12, 18], :width => width-18
            pdf.fill_color "000000"
            pdf.start_new_page unless index == stories.size - 1
          end

          puts Rainbow(">>> Generated PDF file in '#{destination}.pdf'").foreground(:green)
        end
      rescue Exception
        puts Rainbow("[!] There was an error while generating the PDF file... What happened was:").foreground(:red)
        raise
      end

      private

      def setup_font(pdf)
      end

    end
  end
end


lib/pivotal_to_pdf-formatters/large_text.rb
require 'rubygems'
require 'prawn'
require 'rainbow'
require_relative 'default'

module PivotalToPdf
  module Formatters
    class LargeText < Default # same as Default but larger text and no description

      def write_to(destination)
        Prawn::Document.generate("#{destination}.pdf",
                                 :page_layout => :landscape,
                                 :margin => [25, 25, 50, 25],
                                 :page_size => [302, 432]) do |pdf|

          setup_font(pdf)

          stories.each_with_index do |story, index|
            next if story.story_type == 'release' # save some trees

            padding = 10
            pdf.stroke_color = story.story_color
            pdf.stroke_bounds
            width = 370
            pdf.line_width=6
            # --- Write content
            pdf.bounding_box [pdf.bounds.left+padding, pdf.bounds.top-padding], :width => width do
              pdf.text max_len_str(story.formatted_name, 120), :size => 28, :inline_format => true
              pdf.fill_color "52D017"
              pdf.text story.formatted_labels, :size => 16
              pdf.text "\n", :size => 28
              pdf.fill_color "444444"
            end
            pdf.line(pdf.bounds.bottom_left, pdf.bounds.bottom_right)
            pdf.stroke_bounds

            pdf.text_box story.points, :size => 24, :at => [12, 30], :width => width-18 unless story.points.nil?
            pdf.fill_color "999999"
            pdf.text_box story.story_type.capitalize, :size => 16, :align => :right, :at => [12, 18], :width => width-18
            pdf.fill_color "000000"
            pdf.start_new_page unless index == stories.size - 1
          end

          puts Rainbow(">>> Generated PDF file in '#{destination}.pdf'").foreground(:green)
        end
      rescue Exception
        puts Rainbow("[!] There was an error while generating the PDF file... What happened was:").foreground(:red)
        raise
      end

      private

      def max_len_str(str, max_len)
        str.length <= max_len ? str : str[0...max_len-1] + '…'
      end

    end
  end
end
